### PR TITLE
data-encoding: initial integration

### PR DESCRIPTION
**Pull request by DavidKorczynski**
@ia0 would you be open for integrating data-encoding to OSS-Fuzz such that the fuzzers can run continuously?

### CODE REVIEW

File: projects/data-encoding/build.sh [hunk 1]
```diff
@@ -0,0 +1,29 @@
...
+# limitations under the License.
+#
+################################################################################
+
+# Apply diff for lib/fuzz/Cargo.toml to include v3-preview feature for fuzzing
+git apply $SRC/patch.diff
+
+# Build the fuzzers and project source code
+cd lib
+cargo fuzz build -O
```
Comment (ia0) [inline projects/data-encoding/build.sh:23]: You don't really need the patch. You can do:

```suggestion
cargo fuzz build -O --features=data-encoding/v3-preview
```

Comment (ia0) [inline projects/data-encoding/build.sh:23]: What's the rationale to disable debug assertions and overflow checks? Without `-O` we still get a `--release` build, which is where most of the performance will come from. Debug assertions and overflow checks have a pretty high signal to overhead ratio in my opinion.

Here's the documentation of `cargo fuzz build` says:

> By default fuzz targets are built with optimizations equivalent to `cargo build --release`, but with debug assertions and overflow checks enabled. Address Sanitizer is also enabled by default.

File: projects/data-encoding/build.sh [hunk 2]
```diff
@@ -0,0 +1,29 @@
...
+
+# Copy built fuzzer binaries to $OUT
+cp $SRC/data-encoding/target/x86_64-unknown-linux-gnu/release/encoder $OUT
+cp $SRC/data-encoding/target/x86_64-unknown-linux-gnu/release/encode_write $OUT
+cp $SRC/data-encoding/target/x86_64-unknown-linux-gnu/release/round_trip $OUT
+cp $SRC/data-encoding/target/x86_64-unknown-linux-gnu/release/v3-preview $OUT
```
Comment (ia0) [inline projects/data-encoding/build.sh:29]: nit: I'll probably rename, add, and remove fuzzing targets. This should make it more robust.

```suggestion
find $SRC/data-encoding/target/x86_64-unknown-linux-gnu/release -maxdepth 1 \
    -type f -perm -u=x -exec cp {} $OUT \;
```

**Review by ia0 — APPROVED** [review]
Thanks @DavidKorczynski for the suggestion. Yes, I think this is a very good idea. If I can help in any way, feel free to ping me or open an issue on `ia0/data-encoding`.

File: projects/data-encoding/build.sh (hunk 1 above)
Comment (DavidKorczynski) [reply projects/data-encoding/build.sh]: Thanks, fixed!

File: projects/data-encoding/build.sh (hunk 2 above)
Comment (DavidKorczynski) [reply projects/data-encoding/build.sh]: Ah nice, thank you!

**Review by ia0 — APPROVED** [review]
Thanks!

File: projects/data-encoding/build.sh (hunk 1 above)
Comment (DavidKorczynski) [reply projects/data-encoding/build.sh]: Gotcha, no specific reason as such and I agree with your points.

Comment (DavidKorczynski) [reply projects/data-encoding/build.sh]: Switched to no `-O`

**Review by AdamKorcz — APPROVED** [review]

**Comment by ia0** [comment]
By the way, I realized I don't have access to the build status, so I've created #12647 to fix that.

**Pull request merged**